src/Microsoft.Data.SqlClient/src/Microsoft/Data/Sql/SqlDataSourceEnumeratorNativeHelper.cs

```diff
@@ -0,0 +1,182 @@
+                    {
+#if NETFRAMEWORK
+                        readLength = SNINativeMethodWrapper.SNIServerEnumRead(handle, buffer, bufferSize, out more);
+#else
+                        readLength = SNINativeMethodWrapper.SNIServerEnumRead(handle, buffer, bufferSize, out more);
+#endif
+                        SqlClientEventSource.Log.TryTraceEvent("<sc.SqlDataSourceEnumeratorNativeHelper.GetDataSources|INFO> GetDataSources:SNIServerEnumRead returned readlength {0}", readLength);
+                        if (readLength > bufferSize)
+                        {
+                            failure = true;
+                            more = false;
+                        }
+                        else if (0 < readLength)
+                        {
+                            strbldr.Append(buffer);
+                        }
+                    }
+                }
+            }
+            finally
+            {
+                if (handle != ADP.s_ptrZero)
+                {
+                    SNINativeMethodWrapper.SNIServerEnumClose(handle);
+                }
+            }
+
+            if (failure)
+            {
+                Debug.Assert(false, "GetDataSources:SNIServerEnumRead returned bad length");
+                SqlClientEventSource.Log.TryTraceEvent("<sc.SqlDataSourceEnumerator.GetDataSources|ERR> GetDataSources:SNIServerEnumRead returned bad length, requested %d, received %d", bufferSize, readLength);
+                throw ADP.ArgumentOutOfRange("readLength");
+            }
+            return ParseServerEnumString(strbldr.ToString());
+        }
+
+        static private System.Data.DataTable ParseServerEnumString(string serverInstances)
+        {
+            DataTable dataTable = new("SqlDataSources");
+            dataTable.Locale = CultureInfo.InvariantCulture;
+            dataTable.Columns.Add(ServerName, typeof(string));
+            dataTable.Columns.Add(InstanceName, typeof(string));
+            dataTable.Columns.Add(IsClustered, typeof(string));
+            dataTable.Columns.Add(Version, typeof(string));
+            string serverName = null;
+            string instanceName = null;
+            string isClustered = null;
+            string version = null;
+            string[] serverinstanceslist = serverInstances.Split(new string[] { "\0\0\0" }, StringSplitOptions.None);
+            SqlClientEventSource.Log.TryTraceEvent("<sc.SqlDataSourceEnumeratorNativeHelper.ParseServerEnumString|INFO> Number of server instances results recieved are {0}", serverinstanceslist.Length);
+
+            // Every row comes in the format "serverName\instanceName;Clustered:[Yes|No];Version:.." 
+            // Every row is terminated by a null character.
+            // Process one row at a time
+            foreach (string instance in serverinstanceslist)
+            {
+                //  string value = instance.Trim('\0'); // MDAC 91934
+                string value = instance.Replace("\0", "");
```

Comment:
Can we use ReadOnlySpan here for less memory allocation?

Reply:
Replace has been changed by Trim.